### PR TITLE
Add support for SendToGroup/AddGroup/RemoveGroup API and compatible with preview version

### PR DESCRIPTION
**Pull request by JialinXin**
Fix issue #9 , #29

### CODE REVIEW

**Comment by anthonychu** [comment]
@JialinXin I merged #31 hoping to make the diff easier to read, but looks like this caused a small merge conflict. Sorry!

**Comment by JialinXin** [comment]
No worries. I did merge with #31 before and updated tests later which might lead the conflict again. Current changes looks good. Thanks for helping solve this!

File: src/SignalRServiceExtension/Client/AzureSignalRClient.cs [hunk 1]
```diff
@@ -65,13 +66,68 @@ public Task SendToUser(string hubName, string userId, SignalRData data)
...
+            var connectionInfo = GetServerConnectionInfo(hubName, userGroupSegment);
+            var uri = $"{connectionInfo.Url}{userGroupSegment}";
+            return RequestAsync(uri, null, connectionInfo.AccessToken, HttpMethod.Put);
+        }
+
+        public Task RemoveUser(string hubName, string userId, string group)
+        {
+            if (string.IsNullOrEmpty(Version))
+            {
+                throw new ArgumentException($"API AddUserToGroup is support after Version = 1.0, check SignalR connection string to verify Version information");
```
Comment (chenkennt) [inline src/SignalRServiceExtension/Client/AzureSignalRClient.cs]: AddUserToGroup -> RemoveUserToGroup

File: src/SignalRServiceExtension/Client/AzureSignalRClient.cs [hunk 2]
```diff
@@ -65,13 +66,68 @@ public Task SendToUser(string hubName, string userId, SignalRData data)
...
+            {
+                throw new ArgumentException($"{nameof(userId)} cannot be null or empty");
+            }
+            if (string.IsNullOrEmpty(group))
+            {
+                throw new ArgumentException($"{nameof(group)} cannot be null or empty");
+            }
+
+            var userGroupSegment = $"/groups/{group}/users/{userId}";
+            var connectionInfo = GetServerConnectionInfo(hubName, userGroupSegment);
```
Comment (chenkennt) [inline src/SignalRServiceExtension/Client/AzureSignalRClient.cs]: Since you already pass in additionalPath, return server url with additionalPath so you don't need to append it again?

File: src/SignalRServiceExtension/Client/AzureSignalRClient.cs [hunk 3]
```diff
@@ -65,13 +66,68 @@ public Task SendToUser(string hubName, string userId, SignalRData data)
...
+            var uri = $"{connectionInfo.Url}{userGroupSegment}";
+            return RequestAsync(uri, null, connectionInfo.AccessToken, HttpMethod.Delete);
+        }
+
+        private (string EndPoint, string AccessKey, string Version) ParseConnectionString(string connectionString)
```
Comment (chenkennt) [inline src/SignalRServiceExtension/Client/AzureSignalRClient.cs]: This should be a static method.

File: src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                throw new ArgumentException("Unsupport Binding Type.");
+            }
+        }
+
+        public Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return Task.CompletedTask;
+        }
+
+        private string FirstOrDefault(params string[] values)
```
Comment (chenkennt) [inline src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs]: Is this used?

Comment (JialinXin) [reply src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs]: No. Removed.

File: src/SignalRServiceExtension/Client/AzureSignalRClient.cs (hunk 2 above)
Comment (JialinXin) [reply src/SignalRServiceExtension/Client/AzureSignalRClient.cs]: Good catch!

File: src/SignalRServiceExtension/Client/IAzureSignalRSender.cs [hunk 5]
```diff
@@ -10,5 +10,8 @@ internal interface IAzureSignalRSender
     {
         Task SendToAll(string hubName, SignalRData data);
         Task SendToUser(string hubName, string userId, SignalRData data);
+        Task SendToGroup(string hubName, string group, SignalRData data);
+        Task AddUser(string hubName, string userId, string group);
+        Task RemoveUser(string hubName, string userId, string group);
```
Comment (anthonychu) [inline src/SignalRServiceExtension/Client/IAzureSignalRSender.cs]:
```suggestion
        Task AddUserToGroup(string hubName, string userId, string groupName);
        Task RemoveUserFromGroup(string hubName, string userId, string groupName);
```

Consider more descriptive method names. Also `group` -> `groupName` to match other places.

File: src/SignalRServiceExtension/Bindings/SignalRCollectorBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
Comment (anthonychu) [inline src/SignalRServiceExtension/Bindings/SignalRCollectorBuilder.cs]: Missing copyright header

File: src/SignalRServiceExtension/Bindings/SignalRCollectorBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+
+namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
+{
+    internal class SignalRCollectorBuilder<T> : IConverter<SignalRAttribute, IAsyncCollector<T>>
+    {
+        private readonly SignalRConfigProvider _configProvider;
```
Comment (anthonychu) [inline src/SignalRServiceExtension/Bindings/SignalRCollectorBuilder.cs]: Nit: privates in the rest of the codebase don't use `_`. If we want to switch all to this convention, perhaps do it in another commit/PR.

Comment (JialinXin) [reply src/SignalRServiceExtension/Bindings/SignalRCollectorBuilder.cs]: OK. Will uniform styles and might do this in the future.

**Comment by JialinXin** [comment]
I did some tests and will update samples after the PR completes.

**Comment by JialinXin** [comment]
With some tests, looks like JavaScript can’t well support generic class binding type covert. Will send a new commit after resolve this.

**Comment by anthonychu** [comment]
@fabiocav can probably help if you need some guidance.

**Comment by JialinXin** [comment]
@anthonychu I update the sample and you may have a try [here](http://jialinxin.github.io/). In anonymous mode, name is the identity. And simply set default group name as it maybe a little complicated to maintain group information in serverless mode. Function is written in Javascript.

File: src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            this.client = client;
+            this.hubName = hubName;
+            converter = new SignalROutputConverter();
+        }
+
+        public async Task AddAsync(T item, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var convertItem = converter.ConvertToSignalROutput(item);
+
+            if (convertItem.GetType() == typeof(SignalRMessage))
```
Comment (chenkennt) [inline src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs]: Will convertedItem be null?

File: src/SignalRServiceExtension/Bindings/SignalROutputConverter.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            messageConverter = new JObjectToTypeConverter<SignalRMessage>();
+            groupActionConverter = new JObjectToTypeConverter<SignalRGroupAction>();
+        }
+
+        // We accept multiple output binding types and rely on them to determine rest api actions
+        // But in non .NET language, it's not able to convert JObject to different types
+        // So need a converter to accurate convert JObject to either SignalRMessage or SignalRGroupAction
+        public object ConvertToSignalROutput(object input)
+        {
+            if (input.GetType() != typeof(JObject))
```
Comment (chenkennt) [inline src/SignalRServiceExtension/Bindings/SignalROutputConverter.cs]: Same here, will input be null?

File: src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs (hunk 8 above)
Comment (JialinXin) [reply src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs]: Depends on how user pass the object when call us. Let me add null check and throw exceptions.

Comment (JialinXin) [reply src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs]: And convertedItem won't be null because inner exception will throw before goes here.

File: src/SignalRServiceExtension/Bindings/SignalROutputConverter.cs (hunk 9 above)
Comment (JialinXin) [reply src/SignalRServiceExtension/Bindings/SignalROutputConverter.cs]: Leave null check outside.

File: src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        internal SignalRAsyncCollector(IAzureSignalRSender client, string hubName)
+        {
+            this.client = client;
+            this.hubName = hubName;
+            converter = new SignalROutputConverter();
+        }
+
+        public async Task AddAsync(T item, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if(item == null)
```
Comment (chenkennt) [inline src/SignalRServiceExtension/Bindings/SignalRAsyncCollector.cs]: Space after `if`
